File: Easy/2807-InsertGreatestCommonDivisorsinLinkedList.js

// Definition for singly-linked list.
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */

function gcd(a, b) {
    if (b === 0)
        return a
    else
        return gcd(b, (a % b))
}

var insertGreatestCommonDivisors = function (head) {
    if (!head.next) return head;
    let node1 = head;
    let node2 = head.next;
    while (node2) {
        let res = gcd(node1.val, node2.val);
        let gcdNode = new ListNode(res);
        [node1.next, gcdNode.next] = [gcdNode, node2];
        [node1, node2] = [node2, node2.next];
    }
    return head;
};

File: Medium/2807-InsertGreatestCommonDivisorsinLinkedList.c
/**
 * Definition for singly-linked list.
 */
struct ListNode
{
    int val;
    struct ListNode *next;
};

int mcd(int a, int b)
{
    if (a % b == 0)
    {
        return b;
    }
    else
    {
        return mcd(b, a % b);
    }
}

struct ListNode *insertGreatestCommonDivisors(struct ListNode *head)
{
    // if (!head)
    //     return NULL;
    if (!head->next)
        return head;
    struct ListNode *node1 = head;
    struct ListNode *node2 = head->next;
    while (node2)
    {
        int res = mcd(node1->val, node2->val);
        struct ListNode *newNode =
            (struct ListNode *)malloc(sizeof(struct ListNode));
        newNode->val = res;
        node1->next = newNode;
        newNode->next = node2;
        node1 = node2;
        node2 = node2->next;
    }
    return head;
}